### ISSUE TITLE
TRPC support

### ISSUE THREAD

**Issue by drakedeatonuk** [nestjs/nest#10046]
### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe it

The problem is that in my nest apps there's a lack of type safety in our nest apis.

### Describe the solution you'd like

It would be great if TRPC was supported natively within nest, so that I could create typesafe APIs easily. 

### Teachability, documentation, adoption, migration strategy

I think this is potentially a feature that a lot of developers in the nest ecosystem would love.

### What is the motivation / use case for changing the behavior?

Whenever we create new api endpoints, there's ambiguity between nest and our frontend apps because our apis dont have standardised inputs and outputs. If each api endpoint had a type which defined its inputs and outputs, then on the frontend it would be much easier to safely work with the api. This is what TRPC solves and it makes sense if support for it was added to nest as im sure that a growing number of devs will want to use TRPC. 

This feature would make coding easier/quicker/safer for a lot of devs.

**Comment by jmcdo29** [comment]
[I've talked about this quite a bit on Discord already](https://discord.com/channels/520622812742811698/819061892861526016/1002258417068548146)

TL;DR: Right now, we _could _build up an integration very similar to the GraphQL integration, **however**, exposing the TRPC router that the `@trpc/client` application would make use of is the difficult part

**Comment by drakedeatonuk** [comment]
Thanks for the link & info. Cool to hear it's being discussed! I have no idea how you'd solve that problem. 

Would love to see this make it into the nest/angular ecosystem... `rtpc/angular` would also be nice.

**Comment by BrunnerLivio** [comment]
I tried some experiments though I, unfortunately, have my doubts that this will be possible. Or at least with the approach, I had in mind. My API idea was the following:

```typescript
// (pseudo-code)

import { Query, TrpcRouter, Input } from '@nestjs/trpc';

class AppDto {
  @IsEmail()
  email: string;
}

@TrpcRouter()
export class AppTrpc {
  @Query()
  async getUser(@Input() req: AppDto) {
    return `hello ${req.email}`;
  }

  static trpc = getTrpcRouter(???);
}

export type AppRouter = typeof AppTrpc.trpc;
```


The goal of `getTrpcRouter()` should be to return the following construct:

```typescript
function getTrpcRouter(instance: any) {
   // Do some magic...

   return rpc.router().query('getUser', {
    input: z.object({
      email: z.string().email()
    }),
    async resolve(req) {
      return `hello ${req.email}`;
    },
  });
}
```

The crux here is that `static trpc = getTrpcRouter(???)` **needs to be static** in order to work with `typeof ...`. So we can not pass in `static trpc = getTrpcRouter(this)` -- since it's static. Because there is no instance we can not get decorator information.

I may be wrong -- but I think it's virtually impossible to do it this way :( Without the type definitions tRPC becomes virtually useless...

**Comment by jmcdo29** [comment]
@BrunnerLivio This is pretty much the same road block I came to in the mental models I was trying to build around it

**Comment by kamilmysliwiec** [comment]
Thanks for your suggestion!

There are no plans to implement it in the foreseeable future.

If you think your request could live outside Nest's scope, we'd encourage you to collaborate with the [community](https://discord.gg/G7Qnnhy) on publishing it as an open source package.
